engineering-plans/0002-subgraph-schema-merging.md

```diff
@@ -0,0 +1,214 @@
+
+<dl>
+  <dt>Author</dt>
+  <dd>Jorge Olivero</dd>
+
+  <dt>Implements</dt>
+  <dd><a href="../rfcs/0001-subgraph-composition.md">RFC-0001 Subgraph Composition</a></dd>
+
+  <dt>Engineering Plan pull request</dt>
+  <dd><a href="https://github.com/graphprotocol/rfcs/pull/5">Engineering Plan PR</a></dd>
+
+  <dt>Obsoletes (if applicable)</dt>
+  <dd>-</dd>
+
+  <dt>Date of submission</dt>
+  <dd>2019-12-09</dd>
+
+  <dt>Date of approval</dt>
+  <dd>TBD</dd>
+
+  <dt>Approved by</dt>
+  <dd>TBD</dd>
+</dl>
+
+## Summary
+
+Subgraph composition allows a subgraph to import types from another subgraph. Imports are designed to be loosely coupled throughout the deployment and indexing phases, and tightly coupled during query execution for the subgraph.
+
+To generate an API schema for the subgraph, the subgraph schema needs to be merged with the imported subgraph schemas. The merging process needs to take the following into account:
+
+1. An imported schema not being available on the Graph Node
+2. An imported schema mising a type imported by the subgraph schema
+3. A schema imported by subgraph name changes
+4. Ability to tell which subgraph/schema each type belongs to
+
+## Implementation
+
+The schema merging implementation consists of two parts:
+
+1. A cache of subgraph schemas
+2. The schema merging logic
+
+### Schema Merging
+
+Add a `merged_schema(&Schema, HashMap<SchemaReference, Arc<Schema>) -> Schema` function to the `graphql::schema` crate which will add each of the imported types to the provided document with a @subgraphId diretive denoting which subgraph the type came from.
+
+The `HashMap<SchemaReference, Arc<Schema>` includes all of the schemas in the subgraph's import graph which are available on the Graph Node. For each `@import` directive on the subgraph, find the imported types by tracing their path along the import graph.
+
+- If any schema node along that path is missing or if the type is missing in the schema, add a type definition to the subgraphs merged schema with the proper name, a `@subgraphId("...")` directive (if available), and a `@placeholder` directive denoting that type was not found. 
+- If the type is found, copy it and add a `@subgraphId("...")` directive.
+- If the type is imported with the `{ name: "", as: "" }` format, the merged type will include an `@originalName("...")` directive preserving the type name from the original schema.
+
+The `api_schema` function will add all the necessary types and fields for the imported types without requiring any changes.
+
+#### Example #1: Complete merge
+
+Local schema before calling `merged_schema`:
+
+```graphql
+type _Schema_
+  @import(
+    types: ["B"],
+    from: { id: "X" }
+  )
+
+type A @entity {
+  id: ID!
+  foo: B!
+}
+```
+
+Imported Schema X:
+
+```graphql
+type B @entity {
+  id: ID!
+  bar: String
+}
+```
+
+Schema after calling `merged_schema`:
+
+```graphql
+type A @entity {
+  id: ID!
+  foo: B!
+}
+
+type B @entity @subgraphId("X") {
```

Comment:
It's actually `@subgraphId(id: "X")`, because directives require named arguments. You can tell because it breaks the syntax highlighting from here on. 😉